src/make_zero.jl

```diff
@@ -1,404 +1,341 @@
-
-@inline function EnzymeCore.make_zero(x::FT)::FT where {FT<:AbstractFloat}
-    return Base.zero(x)
-end
-@inline function EnzymeCore.make_zero(x::Complex{FT})::Complex{FT} where {FT<:AbstractFloat}
-    return Base.zero(x)
-end
-@inline function EnzymeCore.make_zero(
-    x::Array{FT,N},
-)::Array{FT,N} where {FT<:AbstractFloat,N}
-    return Base.zero(x)
-end
-@inline function EnzymeCore.make_zero(
-    x::Array{Complex{FT},N},
-)::Array{Complex{FT},N} where {FT<:AbstractFloat,N}
-    return Base.zero(x)
-end
-
-@inline function EnzymeCore.make_zero(
-    ::Type{Array{FT,N}},
+"""
+    recursive_map(
+        ::Type{T},
+        f,
+        seen::IdDict,
+        xs::NTuple{N,T},
+        ::Val{copy_if_inactive}=Val(false),
+        isleaftype=Returns(false),
+    ) where {T,N,copy_if_inactive}
+
+Recursively map `f` over the differentiable contents of `N` objects `xs = (x1, x2, ..., xN)`
+of arbitrary but identical type and layout. Returns a new object `y` of the same type and
+layout such that each differentiable leaf value `yi` in `y` equals
+`yi = f(x1i, x2i, ..., xNi)` where `x1i, x2i, ..., xNi` are the corresponding leaf values in
+the `xs`.
+
+For each subtree in the `xs` that can be proven by type to only contain non-differentiable
+values, `f` is not invoked and the corresponding subtree `yi` in `y` is taken from the first
+element of `xs` such that `yi == x1i`. If `copy_if_inactive == false`, this is done by
+sharing, `yi = x1i`; if `copy_if_inactive == true`, it is done by copying,
+`yi = deepcopy(x1i)`.
+
+Each element in `xs` is assumed to have the same structure as `x1 = first(xs)`, including
+which fields, if any, reference the same memory or are undefined. This structure will be
+mirrored in the return value `y`. If this assumption does not hold, errors or incorrect
+results may occur.
+
+A function `isleaftype` can be provided to customize which types are considered leafs:
+values of type `T` such that `isleaftype(T) == true` are not recursed into, but instead
+passed to `f`. Non-array types with `fieldcount(T) == 0`, including built-in floats and
+other primitve types, are always considered leaf types as they cannot be recursed into.
+"""
+@inline function recursive_map(
+    ::Type{T},
+    f::F,
     seen::IdDict,
-    prev::Array{FT,N},
-    ::Val{copy_if_inactive} = Val(false),
-)::Array{FT,N} where {copy_if_inactive,FT<:AbstractFloat,N}
-    if haskey(seen, prev)
-        return seen[prev]
+    xs::NTuple{N,T},
+    ::Val{copy_if_inactive}=Val(false),
+    isleaftype::L=Returns(false),
```

Comment:
Wondering whether this is necessary or if the leaf types could just be hardcoded to `Union{_RealOrComplexFloat,Array{<:RealOrComplexFloat}}`. I'll make a prototype of the vector space wrapper and the updated QuadGK rules to see if customizable leaf types comes in handy.